extensions/functions_datetime.yaml

```diff
@@ -32,16 +36,39 @@ scalar_functions:
       * UNIX_TIME Return number of seconds that have elapsed since 1970-01-01 00:00:00 UTC, ignoring leap seconds.
       * TIMEZONE_OFFSET Return number of seconds of timezone offset to UTC.
 
-      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
-      Examples: "Pacific/Marquesas", "Etc/GMT+1".
-      If timezone is invalid an error is thrown.
+      The range of values returned for QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK, SUNDAY_DAY_OF_WEEK,
+      MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK depends on whether counting starts at 1 or 0. This is governed
+      by the indexing option.
+
+      When indexing is ONE:
+       * QUARTER returns values in range 1-4
+       * MONTH returns values in range 1-12
+       * DAY returns values in range 1-31
+       * DAY_OF_YEAR returns values in range 1-366
+       * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 1-7
+       * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 1-53
+
+      When indexing is ZERO:
+       * QUARTER returns values in range 0-3
+       * MONTH returns values in range 0-11
+       * DAY returns values in range 0-30
+       * DAY_OF_YEAR returns values in range 0-365
+       * MONDAY_DAY_OF_WEEK and SUNDAY_DAY_OF_WEEK return values in range 0-6
+       * MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, and US_WEEK return values in range 0-52
+
+       Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
+       Examples: "Pacific/Marquesas", "Etc/GMT+1".
+       If timezone is invalid an error is thrown.
     impls:
       - args:
           - name: component
             options: [ YEAR, ISO_YEAR, US_YEAR, QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                        SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK, HOUR, MINUTE, SECOND,
                        MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME, TIMEZONE_OFFSET ]
             description: The part of the value to extract.
+          - name: indexing
+            options: [ ONE, ZERO ]
+            description: Start counting from 1 or 0.
```

Comment:
Thanks for pointing that out @jorisvandenbossche. I agree with @rok that it would be best to create separate `impls` entries for the cases in which `indexing` has no effect.

Here's my first stab at that. Four implementations that do not require `indexing` followed by three implementations that do require `indexing`. Does this look OK?

```yaml
    impls:
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, HOUR, MINUTE, SECOND,
                       MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME, TIMEZONE_OFFSET ]
            description: The part of the value to extract.
          - name: x
            value: timestamp_tz
          - name: timezone
            description: Timezone string from IANA tzdb.
            value: string
        return: i64
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, HOUR, MINUTE, SECOND,
                       MILLISECOND, MICROSECOND, SUBSECOND, UNIX_TIME ]
            description: The part of the value to extract.
          - name: x
            value: timestamp
        return: i64
      - args:
          - name: component
            options: [ YEAR, ISO_YEAR, US_YEAR, UNIX_TIME ]
            description: The part of the value to extract.
          - name: x
            value: date
        return: i64
      - args:
         - name: component
           options: [ HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND, SUBSECOND ]
           description: The part of the value to extract.
         - name: x
           value: time
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: timestamp_tz
          - name: timezone
            description: Timezone string from IANA tzdb.
            value: string
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: timestamp
        return: i64
      - args:
          - name: component
            options: [ QUARTER, MONTH, DAY, DAY_OF_YEAR, MONDAY_DAY_OF_WEEK,
                       SUNDAY_DAY_OF_WEEK, MONDAY_WEEK, SUNDAY_WEEK, ISO_WEEK, US_WEEK ]
            description: The part of the value to extract.
          - name: indexing
            options: [ ONE, ZERO ]
            description: Start counting from 1 or 0.
          - name: x
            value: date
        return: i64
```

Reply:
Thanks @ianmcook. I pasted your change into the last commit.